[Battery Manager] New module for adding, activating, and deactivating entries in Test Battery

victoriafoing opened the pull request:
### Brief summary of changes

This PR creates a new module 'Battery Manager' which allows users to browse, add, activate, and deactivate entries in the `test_battery` table.

This module includes the following components:
- A **Browse** tab for searching and deactivating entries in the Test Battery
- An **Add** tab for adding and activating entries in the Test Battery
- AJAX calls for checking for duplicate entries, adding entries, deactivating/activating entries, and getting data for dropwdowns 
- SQL patches for permissions and LorisMenu
- Documentation (README, TestPlan, Help)

### This PR improves LORIS by...

Allowing users (e.g. study administrators) to add to and update the `test_battery` table through a front-end interface.

### This resolves issue...

Of having to add to and update the `test_battery` table through the back-end.

### To test this change...

Check out the instructions in the Test Plan

**Note**: The Test Battery table has a lot of NULL values and uses FilterForm.js, which does not properly filter NULL values. Another PR will have to be made to update FilterForm.js.

johnsaigle commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,138 @@
+{
+
+     $form_data = array(
+                   'Test_name'    => isset($_POST['instrument']) ? $_POST['instrument'] : null,
```
These would be better written as 
```php
'Test_name' => $_POST['instrument'] ?? null;`
```


https://www.virendrachandak.com/techtalk/php-7-null-coalesce-operator/

johnsaigle commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,245 @@
+    function _hasAccess()
+    {
+        //create user object
+        $user =& \User::singleton();
```
I don't think you need the & here

johnsaigle commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,138 @@
+ */
+function checkForDuplicate()
+{
+    $db =& Database::singleton();
```
This should use the namespace syntax i.e. `$db = \Database::singleton()`;

johnsaigle commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,149 @@
+function getFormData()
+{
+
+     $instrument = isset($_POST['instrument']) ? $_POST['instrument'] : null;
```
Same here, please use the ?? operator.

ZainVirani commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,368 @@
+              ref="instrumentOrder"
+              required={false}
+              min="0"
+              max="127"
```
this seems arbitrary

PapillonMcGill commented on modules/battery_manager/README.md:
```diff
@@ -0,0 +1,32 @@
+
+## Intended Users
+
+The Battery Manager module is used by study administrators
```
Missing ending dot

PapillonMcGill commented on modules/battery_manager/help/battery_manager.md:
```diff
@@ -0,0 +1,33 @@
+
+Under the “Add” tab, users will be able to add a new entry to the Test Battery.
+Users must specify information about the entry by:
+- Selecting from a number of dropdown menus including Instrument, Stage, Subproject, Visit Label, and First Visit
```
quotes

ridz1208 commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+<?php
+/**
+ * Battery Manager adder
```
`Battery Manager adder` -> `Battery Manager entry adder` ?

ridz1208 commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+        if (isset($value)) {
+            $query .= $connector . $key . " = :" . $key . " ";
+        } else {
+            $query .= $connector . $key . " IS NULL ";
```
if a field is empty this would return false and yet they would be functionally equivalent, you should add a check for the empty case

PapillonMcGill commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,245 @@
+     * @return bool
+     * @throws ConfigurationException
+     */
+    function _hasAccess()
```
should include return type

PapillonMcGill replied:
Still not resolved  
``` function _hasAccess() :Bool```

same for other function

PapillonMcGill replied:
@victoriafoing in the function declaration.

i.e. "function _hasAccess() : bool"
in the case the function accept parameters:
function name(string $param1, int $param2) : string
you can see more at http://php.net/manual/en/migration70.new-features.php and http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration

ridz1208 commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+    );
+
+    // Return JSON representation of duplicate entry if it exists, null otherwise
+    if ($entry) {
```
generally better to be more explicit, `if(is_null(...))` or `if(empty(...))`... there would be les chances of unexpected behaviour

ridz1208 commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,368 @@
+    }
+     // Check if entry the user is trying to add already exists in the table
+     // Pass in callback function to call after check for duplicate
+    this.isDuplicate(formObj, this.giveOptions);
```
since this is on the front end only there is potential to hack it

PapillonMcGill commented on modules/battery_manager/test/TestPlan.md:
```diff
@@ -0,0 +1,85 @@
+### 📄 Overview
+
+Battery Manager module allows users to search for, add, activate, and deactivate entries in the
+Test Battery
```
should add ending dots at the end of every phrases

ridz1208 commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+/**
```
this check seems to be done on the front-end which is okai but should also happen on the back end since the front end JS code can be manipulated by the user !

validation in general is minimally on the backend, max front end and back end but never FE alone

PapillonMcGill commented on modules/battery_manager/test/TestPlan.md:
```diff
@@ -0,0 +1,85 @@
+  2. Battery Manager module should now have **two** tabs (Browse) and (Add)
+  3. The **Deactivate** should be in the data table
+  4. Clicking on Add tab should hide the data table and display a form with the following fields:
+     Instrument, Minimum age, Maximum age, Stage, Subproject, Visit Label, Site, First Visit,
```
quotes

victoriafoing replied:
could you elaborate where I should add quotes?

PapillonMcGill replied:
each field as they are label that are shown on the user interface.

PapillonMcGill commented on modules/battery_manager/test/TestPlan.md:
```diff
@@ -0,0 +1,85 @@
+## Battery Manager Module Test Plan
+
+### 📄 Overview
```
Should not have icon in the page.

ridz1208 commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+if (isset($_GET['action'])) {
+    $action = $_GET['action'];
+    if ($action == "checkForDuplicate") {
+        echo checkForDuplicate();
```
curiousity, why does this line have an `echo` while line 23 doesnt ?

victoriafoing replied:
because checkForDuplicate must return a JSON representation of the duplicate entry to the JS file (or null if there isn't one)

ridz1208 commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+
+    // Make sure new entry is active
+    $form_data['Active'] = 'Y';
+
```
as mentioned above, this should double check that its not a duplicate entry, it can use the function above to do so

ridz1208 commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+                   'CenterID'     => $_POST['forSite'] ?? null,
+                   'firstVisit'   => $_POST['firstVisit'] ?? null,
+                   'instr_order'  => $_POST['instrumentOrder'] ?? null,
+                  );
```
i think all the entries above should be sanitized 

see https://github.com/aces/Loris/pull/3832/files#diff-a2650fffe7f2fdc8d8fd8f988fd27bc4R144

@johnsaigle please confirm

ridz1208 commented on modules/battery_manager/ajax/get_form_data.php:
```diff
@@ -0,0 +1,74 @@
+    return $stageList;
+}
+
+/**
```
this might be an overkill but I LOVEE ITT

Jkat replied:
Haha, this should totally be converted into a Utility function or something

ridz1208 commented on modules/battery_manager/ajax/update_entry.php:
```diff
@@ -0,0 +1,58 @@
+    );
+
+    if (empty($new_entry)) {
+        throw new Exception("Updated entry but could not fetch it");
```
did you mean to throw an exception or simply showerror as you did in add entry ??
eitherway a comment above explaining why you made that choice can be usefull

ridz1208 commented on modules/battery_manager/ajax/update_entry.php:
```diff
@@ -0,0 +1,58 @@
+        "test_battery",
+        array("Active" => $value),
+        array(
+         "ID" => $_POST["ID"],
```
dont trust post

sanitize

johnsaigle replied:
Technically we should be fine here because `update` does prepared statements and everything.

That said it is probably worth doing a simple check (in this case I think using `is_numeric`) and giving a 400 status code if it fails.

ridz1208 commented on modules/battery_manager/ajax/update_entry.php:
```diff
@@ -0,0 +1,58 @@
+ *
+ * @param string $value which sets Active to Y or N
+ *
+ * @throws Exception
```
add more info -> when ? why ?

ridz1208 commented on modules/battery_manager/ajax/update_entry.php:
```diff
@@ -0,0 +1,58 @@
+        )
+    );
+
+    $new_entry = $DB->pselectRow(
```
I'm not sure why you are doing this? why are you reselcting the values ?

ridz1208 commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,245 @@
+        $this->addBasicText('instrumentOrder', 'Instrument Order');
+        $this->addSelect('active', 'Active', $active);
+
+        $this->tpl_data["hasWritePermission"] = $this->hasWritePermission;
```
tpl_data???

ridz1208 commented on modules/battery_manager/ajax/get_form_data.php:
```diff
@@ -0,0 +1,74 @@
+    return json_encode($formData);
+}
+
+/**
```
functions below seem to be duplications from the main class functions, use those instead, avoid codeduplications

HenriRabalais replied:
@ridz1208 Please elaborate/be more specific about which functions.

ridz1208 commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,245 @@
+
+        // set the class variables
+        $this->columns
+            = [
```
consensus is to user `array(...)` if you found this in another class, the other class should be amended in a separate PR

ridz1208 commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,245 @@
+
+        $this->group_by = '';
+        $this->order_by = 'Full_name';
+        $this->headers  = [
```
`[...]` -> `array(...)`

ridz1208 commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,245 @@
+        $deps    = parent::getCSSDependencies();
+        return array_merge(
+            $deps,
+            [$baseURL . "/battery_manager/css/battery_manager.css"]
```
`[...]` -> `array(...)`

ridz1208 commented:
@davidblader please review JSX/JS code and see if its up to standards

llevitis commented on SQL/Archive/2018_07_23_battery_manager_permissions.sql:
```diff
@@ -0,0 +1,14 @@
+INSERT INTO user_perm_rel (userID, permID) VALUES ((SELECT ID FROM users WHERE UserID = 'admin'), (SELECT permID FROM permissions WHERE code = 'battery_manager_view'));
+
+-- Give edit permission to admin
+INSERT INTO user_perm_rel (userID, permID) VALUES ((SELECT ID FROM users WHERE UserID = 'admin'), (SELECT permID FROM permissions WHERE code = 'battery_manager_edit'));
```
There's no `UserID` equal to "admin" in the IBIS production `users` table, so this patch fails as a NULL value can't be inserted into the `userID` column of `user_perm_rel`.

Jkat replied:
The 'upgrading your loris' wiki section uses ID 1 for the 'admin' user. Until there's a better solution, maybe we could continue with this convention.

johnsaigle commented:
@ridz1208 You'll need to find someone to adopt this PR as @victoriafoing is moving away!

davidblader commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -0,0 +1,137 @@
+    $query     = " SELECT * FROM test_battery ";
+    $i         = 0;
+    $connector = "WHERE ";
+    foreach ($form_data as $key => $value) {
```
not sure about this. in my experience, an entry should be considered a duplicate if there is an existing entry based on `Test_name, Stage, SubprojectID, Visit_label`, some projects might also include `CenterID` as well. 
this would allow duplicates through variances in AgeMin/Max and instr_order

davidblader commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,368 @@
+   */
+  giveOptions(duplicateEntry) {
+        // if duplicate entry exists, convert to JSON
+    if (Object.keys(duplicateEntry).length > 0) {
```
`duplicateEntry` is a string when you pass it to `Object.keys`
i think some other `isset`-esque check would be more appropriate

davidblader commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,368 @@
+      let duplicateEntryJSON = JSON.parse(duplicateEntry);
+          // create object with ID of duplicate entry
+      let entryID =	duplicateEntryJSON.ID;
+      var idObj = new FormData();
```
maybe only define this within the anon function you have within the swal. also use `let`

johnsaigle commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -44,7 +46,7 @@ function checkForDuplicate()
             $connector = "AND ";
         }
-        if (isset($value)) {
+        if (isset($value) || $value !== null) {
```
@victoriafoing I think @ridz1208 wanted you to check for an empty string `""` here instead of null.

johnsaigle replied:
```php
if (isset($value) && !empty($value)) {
```

johnsaigle commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -58,11 +60,18 @@ function checkForDuplicate()
+    
+   
+   // Return JSON representation of duplicate entry if it exists, null otherwise
+    //if ($entry) {
```
These should be deleted if no longer needed. :+1:

victoriafoing commented:
NOTE: I am leaving today but I have made the following changes to the module:
- Changed **Deactivate** column to **Change Status** column and added an **Activate** button in that column
- Added a **Edit Metadata** column with link "Edit" that users can click to edit the entry (similar to Media)

I will spend this week cleaning up my code and responding to the requested changes.
If the module needs more work, I will pass it on to someone else! :)

PapillonMcGill commented on modules/battery_manager/ajax/add_entry.php:
```diff
@@ -116,6 +161,15 @@ function getFormData()
+     
+
+     if (isset($_POST['active'])) {
+         $form_data['Active'] = $_POST['active']; 
```
need to be sanetized

PapillonMcGill commented on modules/battery_manager/ajax/get_form_data.php:
```diff
@@ -18,16 +18,53 @@
     if ($action === "getFormData") {
-        echo getFormData();
+       if (isset($_GET['ID'])) {
+        $entryID = $_GET['ID'];
```
also need to be sanitized

PapillonMcGill commented on modules/battery_manager/ajax/get_form_data.php:
```diff
@@ -18,16 +18,53 @@
+{
+ $db = \Database::singleton();
+
+ $query = " SELECT * FROM test_battery WHERE id = :id ";
```
should specific the list of attributes instead of **\***

victoriafoing commented on modules/battery_manager/test/TestPlan.md:
```diff
@@ -0,0 +1,85 @@
+Battery Manager module allows users to search for, add, activate, and deactivate entries in the
+Test Battery
+
+### 🔒 Permissions 
```
could you rephrase your comment? i'm not sure i understand

PapillonMcGill commented on modules/battery_manager/test/TestPlan.md:
```diff
@@ -1,85 +1,83 @@
-  3. Click on 👉  **column headers**  to make sure sorting functionality is working as expected (Ascending/Descending)
+  1. After a couple of entries are added, make sure they are properly displayed in the data table.
+  2. Make sure that information in the data table corresponds to the information in the `test_battery` table.
+  3. Click on 👉  **column headers**  to make sure sorting functionality is working as expected (Ascending/Descending).
```
icon!

PapillonMcGill commented on modules/battery_manager/ajax/add_or_edit_entry.php:
```diff
@@ -0,0 +1,207 @@
+                  'firstVisit'   => $_POST['firstVisit'] ?? null,
+                  'instr_order'  => $_POST['instrumentOrder'] ?? null,
+                 );
+
```
Data need to be sanitized to prevent SQL injection. Had been flagged before but still not resolved.

driusan replied:
It would be much more robust to prevent SQL injection via prepared statements

PapillonMcGill commented on modules/battery_manager/help/battery_manager.md:
```diff
@@ -0,0 +1,63 @@
+Under the `Browse` tab, use the `Selection Filters` to search for entries by fields such 
+as `Instrument`, `Minimum age (days)`, `Maximum age (days)`, `Stage`, `Subproject`, `Visit Label`,
+`Site`, `First Visit`, `Instrument Order`, and `Active`. As filters are selected, the data table below will
+dynamically update with relevant results. Click the “Clear Filters”
```
suggest to put `Clear Filters` in bold instead than between quotes to be as other references to button on this page.

PapillonMcGill commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,247 @@
+        $firstVisit = $this->getYesNoList();
+
+        $active = $this->getYesNoList();
+
```
section hard to read. Suggest removing empty line and aligning equal signs

PapillonMcGill commented on modules/battery_manager/php/edit.class.inc:
```diff
@@ -0,0 +1,112 @@
+     *
+     * @return none
+     */
+    function setup()
```
please add return type in function declaration

victoriafoing replied:
is "@return void" correct?

victoriafoing replied:
or should i change it to "@return bool" and add a "return true" at the end?

PapillonMcGill replied:
setup() : void

PapillonMcGill commented on modules/battery_manager/ajax/add_or_edit_entry.php:
```diff
@@ -0,0 +1,207 @@
+ *
+ * @return json object or null
+ */
+function checkForDuplicate()
```
please include return type in function definition

driusan replied:
also, the return type appears to be either null or string, not a json object.

driusan commented on modules/battery_manager/README.md:
```diff
@@ -0,0 +1,53 @@
+### Browse
+
+The **browse** functionality is under the `Browse` tab.
+Users can browse entries in the Test Battery and filter the entries using the `Selection Filters`.
```
These paragraphs with line breaks after every sentence are weird and hard to read (also, some of them are too long. `fmt`.

driusan commented on modules/battery_manager/README.md:
```diff
@@ -0,0 +1,53 @@
+
+The **activate/deactivate** functionalities can be accessed in the Menu Table under the `Browse` tab.
+
+In the `Change Status` column, users can directly activate or deactivate an entry by pressing the **Activate** or **Deactivate** button.
```
Is there any protection/check/confirmation/something to ensure they don't accidentally press them?

driusan commented on modules/battery_manager/README.md:
```diff
@@ -0,0 +1,53 @@
+
+In the `Change Status` column, users can directly activate or deactivate an entry by pressing the **Activate** or **Deactivate** button.
+
+In the `Edit Metadata` column, users can indirectly activate or deactivate an entry by navigating to the `Edit` form and changing the value of `Active`. Within an entry's `Edit` form, users can also indirectly activate or deactivate another entry if they edit an entry to the point where the entry becomes the same as another entry in the Test Battery but has a different `Active` value. In this cas ...
```
This functionality seems a little weird and I'm not sure of the reasoning behind it.

Also, is there history kept somewhere of the edits?

Given the below paragraph, maybe it would make more sense to implement edit as "Create new test battery and deactivate the current one." so that nothing is ever lost.

PapillonMcGill replied:
Not sure about the requirement but I would say permit edit to some field and still create + deactivate.

driusan commented on modules/battery_manager/README.md:
```diff
@@ -0,0 +1,53 @@
+
+In the `Edit Metadata` column, users can indirectly activate or deactivate an entry by navigating to the `Edit` form and changing the value of `Active`. Within an entry's `Edit` form, users can also indirectly activate or deactivate another entry if they edit an entry to the point where the entry becomes the same as another entry in the Test Battery but has a different `Active` value. In this cas ...
+
+The **activate** functionality is also in the `Add` tab. If users add an entry that already exists in the Test Battery table but has been deactivated, they will be given the option to activate the existing entry.
```
👍

driusan commented on modules/battery_manager/ajax/add_or_edit_entry.php:
```diff
@@ -0,0 +1,207 @@
+<?php
```
This will probably have to go into another pass of the module but:

1. Add and edit should be two different endpoints, since they're doing two different things
2. They should be implemented as handlers in the php directory, the ajax directory should be deprecated.

driusan commented on modules/battery_manager/ajax/add_or_edit_entry.php:
```diff
@@ -0,0 +1,207 @@
+    } else if ($action == "add") {
+        addEntry();
+    } else if ($action == "edit") {
+        editEntry();
```
Add and edit shouldn't be done via a get request, they should be done via a post or put

Jkat replied:
I concur

driusan commented on modules/battery_manager/ajax/add_or_edit_entry.php:
```diff
@@ -0,0 +1,207 @@
+            $connector = "AND ";
+        }
+        if (isset($value) || $value !== null) {
+            $query .= $connector . $key . " = :" . $key . " ";
```
This is dangerous and should be using a prepared statement.

victoriafoing replied:
to find a duplicate of an entry, we have to build an SQL query that looks at all the values of the entry. the for loop goes through all the values in the entry and determines if they have been filled out or not. if a value is filled out it adds " $key = :$key ". if a value is not filled out, it is NULL so the SQL query must contain " $key IS NULL " instead of " $key = :$key " and the $key must be removed from the parameter array.
for example, if I add an entry with the following values:
Test_name: bmi, AgeMinDays: 0, AgeMaxDays: 99999, Stage: Visit, and I do not fill out SubprojectID, Visit_label, CenterID, firstVisit, instr_order. 
The SQL query is "SELECT * FROM test_battery WHERE Test_name = :Test_name AND AgeMinDays =:AgeMinDays AND AgeMaxDays = :AgeMaxDays AND Stage = :Stage AND SubprojectID IS NULL AND Visit_label IS NULL AND CenterID is NULL AND firstVisit IS NULL AND instr_order IS NULL" with parameters ['Test_name' => 'bmi', 'AgeMinDays' => '0', 'AgeMaxDays' => '99999', 'Stage' => 'Visit']

driusan commented on modules/battery_manager/ajax/add_or_edit_entry.php:
```diff
@@ -0,0 +1,207 @@
+        $message = 'An unknown error occurred!';
+    }
+    header('HTTP/1.1 500 Internal Server Error');
+    header('Content-Type: application/json; charset=UTF-8');
```
This doesn't seem to be return json? (or doing anything with message?)

victoriafoing replied:
if i add exit($message) in the last line, the error message is passed to the front-end. however, i was told to avoid die/exit. is there a better to return the error?

driusan commented on modules/battery_manager/ajax/change_active_status.php:
```diff
@@ -0,0 +1,63 @@
+    );
+
+    $new_entry = $DB->pselectRow(
+        " SELECT * FROM test_battery WHERE ID = :batteryID AND Active = :active",
```
Is there a reason this can't be an explicit list of columns instead of `*`?

driusan commented on modules/battery_manager/help/battery_manager.md:
```diff
@@ -0,0 +1,63 @@
+# Battery Manager
```
The line lengths make this really hard to read too.

driusan commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,357 @@
+
+  render() {
+    // Data loading error
+    if (this.state.error !== undefined) {
```
If `this.state.error` is the empty string, this will show up, which I don't think is intended.

driusan commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,357 @@
+    }
+
+    // Waiting for data to load
+    if (!this.state.isLoaded) {
```
Can this use @HenriRabalais's new loader component?

zaliqarosli replied:
@driusan doesn't it need to be merged first before we can use it?

driusan replied:
good point.. I thought it was merged, ignore my comment until it is.

intralizee commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,357 @@
+  }
+
+  componentDidMount() {
+    var self = this;
```
let self = this;

intralizee commented on modules/battery_manager/jsx/addEntry.js:
```diff
@@ -0,0 +1,357 @@
+    }
+
+    // Inform users about duplicate entries
+    var helpText = (
```
const helpText = ...

intralizee commented on modules/battery_manager/jsx/batteryManagerIndex.js:
```diff
@@ -0,0 +1,260 @@
+   */
+  formatColumn(column, cell, rowData, rowHeaders) {
+     // Create the mapping between rowHeaders and rowData in a row object.
+    var row = {};
```
let row = {};

intralizee commented on modules/battery_manager/jsx/batteryManagerIndex.js:
```diff
@@ -0,0 +1,260 @@
+     // Create Change Status column with Active and Deactivate buttons
+    if (column === 'Change Status') {
+      let entryID = row['Change Status'];
+      var idObj = new FormData();
```
let idObj = new FormData();

intralizee commented on modules/battery_manager/jsx/batteryManagerIndex.js:
```diff
@@ -0,0 +1,260 @@
+    // Create Edit Metadata column with Edit link that allows user to edit entry
+    if (column === 'Edit Metadata') {
+      let entryID = row['Edit Metadata'];
+      var editURL = loris.BaseURL + "/battery_manager/edit/?id=" + entryID;
```
const editURL = ...

zaliqarosli commented on modules/battery_manager/php/edit.class.inc:
```diff
@@ -0,0 +1,112 @@
+ */
+class Edit extends \NDB_Form
+{
+    public $skipTemplate = true;
```
do we need this line if we are not using smarty templates?

driusan replied:
the line is explicitly telling it not to use smarty templates.

zaliqarosli replied:
but is it required if there are no smart templates in the module?

victoriafoing replied:
i think its necessary because when i comment out this line, i get:
"The following errors occurred while attempting to display this page:
 Unable to load template 'file:menu_battery_manager.tpl'"

zaliqarosli commented on modules/battery_manager/php/edit.class.inc:
```diff
@@ -0,0 +1,112 @@
+     * @return bool
+     * @throws ConfigurationException
+     */
+    function _hasAccess()
```
not a big deal, but let's put hasAccess at the top of this class

zaliqarosli commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,247 @@
+{
+    public $hasWritePermission = false;
+    public $AjaxModule         = true;
+    public $skipTemplate       = true;
```
not sure if this is needed is there are no smarty templates

HenriRabalais replied:
The page will not load unless `$skipTemplate` is set to `true`

zaliqarosli commented on modules/battery_manager/jsx/editEntryIndex.js:
```diff
@@ -0,0 +1,30 @@
+ * @author Victoria Foing
+ *
+ */
+import BatteryManagerEditForm from './editEntry';
```
1. it seems to me that there should really be only 1 index file that renders all the other components. 
2. if this index file only imports 1 component (editEntry), then I believe the call to render can just be appended to the end of the editEntry class

Jkat commented:
Hi, don't forget to add this module to the default `0000-00-02-Menus.sql` file

Jkat commented on modules/battery_manager/php/battery_manager.class.inc:
```diff
@@ -0,0 +1,247 @@
+        return array_merge(
+            $deps,
+            array(
+             $baseURL . "/battery_manager/css/battery_manager.css",
```
This part may not be needed if the .css file committed is empty

Jkat commented on modules/battery_manager/php/edit.class.inc:
```diff
@@ -0,0 +1,112 @@
+        return array_merge(
+            $deps,
+            array(
+             $baseURL . "/battery_manager/css/battery_manager.css",
```
This part may not be needed if the .css file committed is empty

victoriafoing commented:
Hi everyone! Thanks for the feedback. I have tackled as much of it as I can.
This PR is now ready to be passed on to someone else.

Cheers,
Vicky :)

PapillonMcGill commented on modules/battery_manager/README.md:
```diff
@@ -2,16 +2,17 @@
 
-The Battery Manager module allows users to **browse**, **add**, **edit**, and **activate/deactivate** entries 
-in the Test Battery. 
+The Battery Manager module allows users to **browse**, **add**, **edit**, **activate**, **deactivate** entries 
```
should be ``` ... activate and deactivate ...```

PapillonMcGill commented on modules/battery_manager/README.md:
```diff
@@ -22,32 +23,52 @@ Users can browse entries in the Test Battery and filter the entries using the `S
+### Activate
 
-The **activate/deactivate** functionalities can be accessed in the Menu Table under the `Browse` tab.
+The **activate** functionality can be accessed in the `Browse` tab, `Edit` window, and the `Add` tab..
```
1 extra dot at the end

HenriRabalais commented on SQL/Archive/2018_07_23_battery_manager_permissions.sql:
```diff
@@ -0,0 +1,14 @@
+-- Add Battery Manager to LorisMenu
```
`SQL/New_patches` doesn't seem to exist, should I put it there anyways?

HenriRabalais replied:
@ridz1208 Please answer.

HenriRabalais commented on SQL/Archive/2018_07_23_battery_manager_permissions.sql:
```diff
@@ -0,0 +1,14 @@
+-- Add Battery Manager to LorisMenu
+INSERT INTO LorisMenu (Parent, Label, Link, Visible, OrderNumber) VALUES (6, 'Battery Manager', 'battery_manager/', null, 7);
```
Could you specify exactly what should be selected for? Is it only the `Parent`?

e.g. (SELECT `Parent` From `LorisMenu` WHERE `Parent` = `Admin`)

HenriRabalais commented on SQL/Archive/2018_07_23_battery_manager_permissions.sql:
```diff
@@ -0,0 +1,14 @@
+INSERT INTO permissions (code, description, categoryID) VALUES ('battery_manager_edit', 'Add, activate, and deactivate entries in Test Battery', 2);
+
+-- Give view permission to admin
+INSERT INTO user_perm_rel (userID, permID) VALUES ((SELECT ID FROM users WHERE UserID = 'admin'), (SELECT permID FROM permissions WHERE code = 'battery_manager_view'));
```
Could you give a quick example of how that would work in this context?

HenriRabalais replied:
@ridz1208 Please answer.

HenriRabalais commented on modules/battery_manager/help/battery_manager.md:
```diff
@@ -0,0 +1,33 @@
+# Battery Manager
```
Would love to but I'm afraid I have no idea. I will make a note and try and fill it in later.

HenriRabalais commented:
Took over maintenance of PR from @victoriafoing and addressed reviews. Major JSX refactoring to come.

PapillonMcGill commented on SQL/New_patches/2018_07_23_battery_manager_permissions.sql:
```diff
@@ -1,7 +1,8 @@
+INSERT INTO LorisMenu (Parent, Label, Link, Visible, OrderNumber)
        VALUES (
-           (SELECT ID FROM LorisMenu WHERE Label = 'Admin', 'Battery Manager'),
+           (SELECT ID FROM (SELECT * FROM LorisMenu) as lm WHERE Label = 'Admin'),
```
... (SELECT ID, 'Battery Manager, null, 7 FROM LorisMenu where Label = 'Admin' );...

HenriRabalais replied:
Instead of `SELECT *`?

PapillonMcGill replied:
instead of lines 4-9

PapillonMcGill replied:
Just saw your updated comment.

Right, you can't insert in the same table your selecting from.

PapillonMcGill replied:
SELECT ID FROM (SELECT ID FROM LorisMenu WHERE Label = 'Admin') AS LM, ...

HenriRabalais replied:
Sounds good!
